small/mempool_malloc.c

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright 2010-2023, Tarantool AUTHORS, please see AUTHORS file.
+ *
+ * Redistribution and use in source and binary forms, with or
+ * without modification, are permitted provided that the following
+ * conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above
+ *    copyright notice, this list of conditions and the
+ *    following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above
+ *    copyright notice, this list of conditions and the following
+ *    disclaimer in the documentation and/or other materials
+ *    provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY <COPYRIGHT HOLDER> ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+ * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+ * <COPYRIGHT HOLDER> OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+ * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+ * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+#include "mempool.h"
+
+enum {
+	/** Maximum alignment for the ASAN mempool allocation. */
+	MEMPOOL_ASAN_MAX_ALIGNMENT = 4096,
+};
+
+void
+mempool_create(struct mempool *pool, struct slab_cache *cache,
+	       uint32_t objsize)
+{
+	(void)cache;
+	small_assert(objsize > 0);
+	pool->objsize = objsize;
+	pool->objcount = 0;
+	rlist_create(&pool->objects);
+
+	/* Alignment is explained in header. */
+	pool->alignment = 1 << (__builtin_ffs(objsize) - 1);
+	if (pool->alignment > MEMPOOL_ASAN_MAX_ALIGNMENT)
+		pool->alignment = MEMPOOL_ASAN_MAX_ALIGNMENT;
```

Comment:
Clarified comment on alignment in header:
```c
/*
 * Allocation alignment is the same as in regular implementation.
 * Alignment is determined by the size of mempool object. If size has
 * 0 bit set then alignment is 1, if size has 1 bit set (with lower bits
 * set to 0) then alignment is 2 etc etc. If size is power of 2 then
 * alignment is same as size. This alignment follows from the fact that
 * regular mempool is just an array with last element aligned on very
 * high power of 2.
 *
 * However for ASAN implementation alignment is limited to
 * MEMPOOL_ASAN_MAX_ALIGNMENT to prevent high memory waste. Typically we
 * don't need high alignment while merely object size may force it to be high.
 *
 * Additionally each allocation is not aligned on next power of 2 alignment.
 * This improves unaligned memory access check.
*/
```